### PR TITLE
Vercel migration

### PR DESCRIPTION
**Pull request by john-rock**
## What are you changing in this pull request and why?

This adds all of the adjustments needed to handle the Netlify to Vercel migration.
- Adjusts serverless functions to support Vercel syntax
- Adds `vercel.json` file to handle redirects
- Adds function to append a query string param to image sources only if hosted through vercel to prevent broken cached images from being shown post migration.

⚠️ Note that this is setup to load on both Netlify and Vercel that way we can merge this PR into `current` and then update the DNS when ready to start on Vercel. ⚠️

## Preview Links
Vercel: https://docs-getdbt-com-git-vercel-migration-dbt-labs.vercel.app
Netlify: https://deploy-preview-3968--docs-getdbt-com.netlify.app

## Testing
In addition to our internal testing documentation, the following steps should be performed and confirmed before merging into `current`.
1. Confirm `get-discourse-comments` function works on both Netlify and Vercel previews. Should render comment section with a test comment.
Vercel: https://docs-getdbt-com-git-vercel-migration-dbt-labs.vercel.app/blog/create-dbt-documentation-10x-faster-with-ChatGPT
Netlify: https://deploy-preview-3968--docs-getdbt-com.netlify.app/blog/create-dbt-documentation-10x-faster-with-chatgpt

2. Confirm `get-discourse-topics` DiscourseFeed component works on both Netlify and Vercel Previews.
Vercel Forum Page: https://docs-getdbt-com-git-vercel-migration-dbt-labs.vercel.app/community/forum
Netlify Forum Page: https://deploy-preview-3968--docs-getdbt-com.netlify.app/community/forum

3. Confirm `get-discourse-topics` DiscourseHelpFeed component renders under "Questions from the Community" on the following pages
Vercel: https://docs-getdbt-com-git-vercel-migration-dbt-labs.vercel.app/docs/build/incremental-models
Netlify: https://deploy-preview-3968--docs-getdbt-com.netlify.app/docs/build/incremental-models

4. Confirm redirects work. Once we are on Vercel all redirects will need to be added in a [vercel.json](https://github.com/dbt-labs/docs.getdbt.com/blob/836e16373a4d20c80f83db1a895c454341210af0/website/vercel.json) file. Until then, if redirects are added they will need to be added in both `_redirects` and `vercel.json`.

## Additional Comments
@JKarlavige in an effort to try and prevent cached images once the migration is complete I created a [new function ](https://github.com/dbt-labs/docs.getdbt.com/blob/836e16373a4d20c80f83db1a895c454341210af0/website/functions/image-cache-wrapper.js)that will append a query param only if the site is served from Vercel that way we don't bust the cache prior to migration. 

Right now it is only applying to image sources used in functional components and not markdown files. I originally was attempting to build a plugin that applied a param to all image sources regardless of where they were set but was having trouble getting it to work. Any thoughts on how we can improve this so we don't have to manually add a param to markdown image sources after migration?

## Checklist
- [x] Confirm Discourse comments and discourse topics load correctly on both Vercel and Netlify deploy previews
- [x] Confirm redirects work
- [x] Confirm Algolia search works

### CODE REVIEW

**Comment by JKarlavige** [comment]
> @JKarlavige in an effort to try and prevent cached images once the migration is complete I created a [new function ](https://github.com/dbt-labs/docs.getdbt.com/blob/836e16373a4d20c80f83db1a895c454341210af0/website/functions/image-cache-wrapper.js)that will append a query param only if the site is served from Vercel that way we don't bust the cache prior to migration.
> 
> Right now it is only applying to image sources used in functional components and not markdown files. I originally was attempting to build a plugin that applied a param to all image sources regardless of where they were set but was having trouble getting it to work. Any thoughts on how we can improve this so we don't have to manually add a param to markdown image sources after migration?

Awesome, good idea setting up a function to add to the end of images. As far as handling markdown files, i'm not 100% sure on how we can do this. I can try to look around to see what options are available, otherwise this might be a good topic for our standing meeting for when you're back from vaca. There has to be a way to prevent caching across the board for this.

One thing that might work in our favor - clicking through the docs site it looks like the images have standard `src` paths, such as `/img/redshift_tutorial/images/cloud_formation_in_progress.png`. If the images throughout docs don't have the netlify optimized paths and the src stays the same, i feel like the images would work?

File: website/api/get-discourse-comments.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+      if (typeof topicId === "number") {
+        comments = await getDiscourseTopicbyID(topicId);
+        comments.shift();
+        comments = { topicId, comments };
+
+        return await response.status(200).json(comments);
+      } else {
+        return await returnResponse(500, {
+          error: "Unable to create Discourse topic TopicID is not a number.",
+        });
```
Comment (JKarlavige) [inline website/api/get-discourse-comments.js]: I believe this needs to be updated to the `response.status().json()` method [suggested by Vercel.](https://vercel.com/docs/functions/serverless-functions/quickstart#create-a-serverless-function)

File: website/api/get-discourse-comments.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+        let topicId = await response.data.topic_id
+
+        console.log('Topic successfully created with topic_id', topicId)
+
+        return topicId
+    
+    } catch(err) {
+        console.log('err on createDiscourseTopic', err)
+        return await returnResponse(500, { error: 'Unable to create Discourse topic.'})
```
Comment (JKarlavige) [inline website/api/get-discourse-comments.js]: Same note to update this to the Vercel response method.

File: website/api/get-discourse-comments.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            }, { headers })
+        }
+
+        // Filter only 'regular' posts in Discourse. (e.g. not moderator actions, small_actions, whispers)
+        post_stream.posts = post_stream.posts.filter(post => post.post_type === 1)
+        
+        return post_stream.posts
+    } catch(err) {
+        console.log('err on getDiscourseTopicbyID', err)
+        return await returnResponse(500, { error: 'Unable to get Discourse topic by ID.'})
```
Comment (JKarlavige) [inline website/api/get-discourse-comments.js]: Same note to update this to the Vercel response method.

File: website/api/get-discourse-comments.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    }
+}
+
+async function searchDiscourseExternalId(externalId) {
+    console.log(`Searching for external_id in Discourse - ${externalId}`);
+    try {
+        const data = await axios.get(`${discourse_endpoint}/t/external_id/${externalId}.json`, { headers });
+        return data.data.id;
+    } catch (err) {
+        return await returnResponse(500, { error: 'Unable to search Discourse topics.' });
```
Comment (JKarlavige) [inline website/api/get-discourse-comments.js]: Same note to update this to the Vercel response method.

File: website/api/get-discourse-comments.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    'Access-Control-Allow-Headers': '*',
+    'Access-Control-Allow-Methods': 'POST, OPTIONS, GET' 
+  }
+  const resObj = {
+    statusCode: status,
+    headers,
+    body: JSON.stringify(res)
+  }
+  return resObj
+}
```
Comment (JKarlavige) [inline website/api/get-discourse-comments.js]: Then if we update all responses we can likely nuke this method

File: website/api/get-discourse-topics.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    // Get topics from Discourse
+    let { data: { posts, topics } } = await axios.get(`${discourse_endpoint}/search?q=${query}`, { headers })
+
+    // Return empty array if no topics found for search query
+    // 200 status is used to prevent triggering Datadog alerts
+    if(!topics || topics?.length <= 0) {
+      // Log message with encoded query and end function
+      console.log('Unable to get results from api request.')
+      console.log(`Search query: ${query}`)
+      return returnResponse(200, [])
```
Comment (JKarlavige) [inline website/api/get-discourse-topics.js]: Same note to update this to the Vercel response method.

File: website/api/get-discourse-topics.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      }, [])
+    }
+
+    // Return topics 
+    //return await returnResponse(200, allTopics)
+    return await response.status(200).json(allTopics)
+  } catch(err) {
+    // Log and return the error
+    console.log('err', err)
+    return await returnResponse(500, { error: 'Unable to get topics from Discourse.'})
```
Comment (JKarlavige) [inline website/api/get-discourse-topics.js]: Same note to update this to the Vercel response method.

File: website/api/get-discourse-topics.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+      headers,
+      body: JSON.stringify(res)
+    }
+
+    return resObj;
+  } catch (error) {
+    console.error('Error:', error); 
+    throw error;
+  }
+}
```
Comment (JKarlavige) [inline website/api/get-discourse-topics.js]: Can nuke

**Review by JKarlavige — CHANGES_REQUESTED** [review]
Overall everything seems to be working great. Just have one note on the api responses, where some are currently using the old netlify responses.

**Review by runleonarun — COMMENTED** [review]
@john-rock and @JKarlavige This is looking great! Here's a Loom that summarizes my findings with preview comments (they're not working):

https://www.loom.com/share/4e28f0dcc98f433d9414daa40422a7a2?sid=282155b7-ecbb-4705-ba16-42654aa422f5

**Comment by john-rock** [comment]
> Overall everything seems to be working great. Just have one note on the api responses, where some are currently using the old netlify responses.

Hey @JKarlavige so I dug into this, and correct me if I'm wrong, but it looks like the Vercel `response` object is only available on the exported serverless function (getDiscourseTopics & getDiscourseComments). I tried using the `response` on the others and it was returning `response.status` as `undefined`.

I did however refactor the error responses on the other functions and was able to remove the `returnResponse` functions for both.

Take a look and let me know if you have concerns.

**Comment by JKarlavige** [comment]
> > Overall everything seems to be working great. Just have one note on the api responses, where some are currently using the old netlify responses.
> 
> Hey @JKarlavige so I dug into this, and correct me if I'm wrong, but it looks like the Vercel `response` object is only available on the exported serverless function (getDiscourseTopics & getDiscourseComments). I tried using the `response` on the others and it was returning `response.status` as `undefined`.
> 
> I did however refactor the error responses on the other functions and was able to remove the `returnResponse` functions for both.
> 
> Take a look and let me know if you have concerns.

Ahh yeah that makes sense, should only required the `response` for Vercel functions. Clicked through multiple pages where the endpoints are being used and all looks to be working correctly! Verified redirects are working and images are loading correctly with the new wrapper. Awesome work!

**Review by JKarlavige — APPROVED** [review]
:shipit: :shipit: :shipit:

**Comment by john-rock** [comment]
Hey @runleonarun thanks for the providing the loom! I looked into it and can't seem to find documentation on a deploy preview comment to Github code comment integration.

The `Vercel Preview Comments` check is defaulted as a non blocking check when deploy preview comments are turned on and is there to warn users that there is deploy preview comments that need to be addressed. After acknowledging the comments using the steps below in the screenshot the check is then switched to successful. 
Python_models___dbt_Developer_Hub

I also was not a part of the Vercel demo so I may not have not have full context into the feature you are asking about. @JKarlavige do you have any more info on this?

It does look like there is a Slack integration that may be similair to the feature you menitoned but that currently looks like it is in beta 🤔. But even then it only looks like it will show the actual comments not where it is located in the code.
https://vercel.com/docs/workflow-collaboration/comments/integrations

**Comment by JKarlavige** [comment]
> I also was not a part of the Vercel demo so I may not have not have full context into the feature you are asking about. @JKarlavige do you have any more info on this?
> 
> It does look like there is a Slack integration that may be similair to the feature you menitoned but that currently looks like it is in beta 🤔. But even then it only looks like it will show the actual comments not where it is located in the code. https://vercel.com/docs/workflow-collaboration/comments/integrations

I believe the Preview Comments check is working correctly here. If review comments are made with vercel, that check will fail until all comments are resolved. I haven't played around with that much, but i believe that is the expected functionality.

@john-rock did you receive a notification that comments were made using the vercel preview feature? As long as we're getting notifications i believe it should be good to go

**Comment by john-rock** [comment]
> @john-rock did you receive a notification that comments were made using the vercel preview feature? As long as we're getting notifications i believe it should be good to go

Yep I got an email on every new comment and looks like that is the only available notification method at the moment https://vercel.com/docs/workflow-collaboration/comments/how-comments-work

> Pull request owners receive emails whenever a new comment is created. Comment creators and participators in comment threads will receive email notifications alerting them to new activity within those threads.

**Review by runleonarun — APPROVED** [review]
@john-rock and @JKarlavige that's too bad they don't turn into code comments, but otherwise this looks good form docs POV!

I am curious about the Slack integration.

**Comment by john-rock** [comment]
> @john-rock and @JKarlavige that's too bad they don't turn into code comments, but otherwise this looks good form docs POV!
> 
> I am curious about the Slack integration.

[It looks like](https://vercel.com/integrations/slack-beta) it removes the step from needing to open the inbox pane on the deploy preview and have the comments automatically sync to a channel with the ability to comment back and resolve the comment directly in the slack thread.

If that is something that would improve the workflow for the docs team we should explore it more once we have these Vercel migrations complete!

**Pull request merged**